compile error

/usr/bin/ld: CMakeFiles/symtabAPI.dir/src/Object.C.o: in function `tbb::cache_aligned_allocator<tbb::interface5::internal::hash_map_base::bucket>::deallocate(tbb::interface5::internal::hash_map_base::bucket*, unsigned long)':
/usr/include/tbb/cache_aligned_allocator.h:88: undefined reference to `tbb::internal::NFS_Free(void*)'
/usr/bin/ld: /usr/include/tbb/cache_aligned_allocator.h:88: undefined reference to `tbb::internal::NFS_Free(void*)'
/usr/bin/ld: /usr/include/tbb/cache_aligned_allocator.h:88: undefined reference to `tbb::internal::NFS_Free(void*)'
/usr/bin/ld: CMakeFiles/symtabAPI.dir/src/Object.C.o: in function `tbb::spin_rw_mutex_v3::scoped_lock::acquire(tbb::spin_rw_mutex_v3&, bool)':
/usr/include/tbb/spin_rw_mutex.h:118: undefined reference to `tbb::spin_rw_mutex_v3::internal_acquire_reader()'
/usr/bin/ld: CMakeFiles/symtabAPI.dir/src/Object.C.o: in function `tbb::spin_rw_mutex_v3::scoped_lock::upgrade_to_writer()':
/usr/include/tbb/spin_rw_mutex.h:127: undefined reference to `tbb::spin_rw_mutex_v3::internal_upgrade()'
/usr/bin/ld: CMakeFiles/symtabAPI.dir/src/Object.C.o: in function `tbb::spin_rw_mutex_v3::scoped_lock::try_acquire(tbb::spin_rw_mutex_v3&, bool)':
/usr/include/tbb/spin_rw_mutex.h:162: undefined reference to `tbb::spin_rw_mutex_v3::internal_try_acquire_writer()'

--- a/common/h/concurrent.h
+++ b/common/h/concurrent.h
@@ -54,10 +54,10 @@ namespace dyn_c_annotations {
 
 template<typename K, typename V>
 class dyn_c_hash_map : protected tbb::concurrent_hash_map<K, V,
-    tbb::tbb_hash_compare<K>, std::allocator<std::pair<K,V>>> {
+    tbb::tbb_hash_compare<K>, std::allocator<std::pair<const K,V>>> {
 
     typedef tbb::concurrent_hash_map<K, V,
-        tbb::tbb_hash_compare<K>, std::allocator<std::pair<K,V>>> base;
+        tbb::tbb_hash_compare<K>, std::allocator<std::pair<const K,V>>> base;
 public:
     using typename base::value_type;
     using typename base::mapped_type;

--- a/common/h/concurrent.h
+++ b/common/h/concurrent.h
@@ -40,6 +40,7 @@
 #include <tbb/concurrent_hash_map.h>
 #include <tbb/concurrent_vector.h>
 #include <tbb/concurrent_queue.h>
+#include <boost/functional/hash.hpp>
 
 namespace Dyninst {
 
@@ -52,12 +53,54 @@ namespace dyn_c_annotations {
     void COMMON_EXPORT runlock(void*);
 }
 
+namespace concurrent {
+  template <typename K>
+  struct hasher {
+    size_t operator()(K const& k) const {
+      return boost::hash<K>{}(k);
+    }
+  };
+
+  namespace detail {
+    template <bool, typename Key>
+    struct hash_compare;
+
+    // New style tbb_hash_compare concept (TBB_VERSION_MAJOR >= 2021)
+    template<typename Key>
+    class hash_compare<true, Key> {
+    	hasher<Key> my_hasher;
+    public:
+    	size_t hash(Key const& k) const {
+    		return my_hasher(k);
+    	}
+    	bool equal(Key const& k1, Key const& k2) const {
+    		return k1 == k2;
+    	}
+    };
+
+    // Old style tbb_hash_compare concept
+    template<typename Key>
+    class hash_compare<false, Key> {
+    public:
+    	static size_t hash(Key const& k) {
+    		return hasher<Key>{}(k);
+    	}
+    	static bool equal(Key const& k1, Key const& k2) {
+    		return k1 == k2;
+    	}
+    };
+  }
+}
+
 template<typename K, typename V>
 class dyn_c_hash_map : protected tbb::concurrent_hash_map<K, V,
-    tbb::tbb_hash_compare<K>, std::allocator<std::pair<const K,V>>> {
+    concurrent::detail::hash_compare<TBB_VERSION_MAJOR >= 2021, K>,
+	std::allocator<std::pair<const K,V>>> {
 
-    typedef tbb::concurrent_hash_map<K, V,
-        tbb::tbb_hash_compare<K>, std::allocator<std::pair<const K,V>>> base;
+	using base = tbb::concurrent_hash_map<K, V,
+			concurrent::detail::hash_compare<TBB_VERSION_MAJOR >= 2021, K>,
+			std::allocator<std::pair<const K,V>>>;
+
 public:
     using typename base::value_type;
     using typename base::mapped_type;

--- a/symtabAPI/src/dwarfWalker.h
+++ b/symtabAPI/src/dwarfWalker.h
@@ -26,30 +26,28 @@
 #include <bits/stdc++.h>
 
 namespace Dyninst {
-namespace SymtabAPI {
-    typedef struct {
-        Dwarf_Off off;
-        bool file;
-        Module * m;
-    } type_key;
-}
-}
-
-namespace tbb {
-    using namespace Dyninst::SymtabAPI;
-    template<>
-    struct tbb_hash_compare<type_key> {
-        static size_t hash(const type_key& k) {
-            size_t seed = 0;
-            boost::hash_combine(seed, k.off);
-            boost::hash_combine(seed, k.file);
-            boost::hash_combine(seed, static_cast<void *>(k.m));
-            return seed;
-        }
-        static bool equal(const type_key& k1, const type_key& k2) {
-            return (k1.off==k2.off && k1.file==k2.file && k1.m==k2.m);
-        }
-    };
+	namespace SymtabAPI {
+		typedef struct {
+			Dwarf_Off off;
+			bool file;
+			Module * m;
+		} type_key;
+		inline bool operator==(type_key const& k1, type_key const& k2) {
+			return k1.off==k2.off && k1.file==k2.file && k1.m==k2.m;
+		}
+	}
+	namespace concurrent {
+		template<>
+		struct hasher<SymtabAPI::type_key> {
+			size_t operator()(const SymtabAPI::type_key& k) const {
+				size_t seed = 0;
+				boost::hash_combine(seed, k.off);
+				boost::hash_combine(seed, k.file);
+				boost::hash_combine(seed, static_cast<void *>(k.m));
+				return seed;
+			}
+		};
+	}
 }
 
 namespace Dyninst {

